README.md
<p align="center">
    <a href="https://github.com/yii2tech" target="_blank">
        <img src="https://avatars2.githubusercontent.com/u/12951949" height="100px">
    </a>
    <h1 align="center">Spreadsheet Data Export extension for Yii2</h1>
    <br>
</p>

This extension provides ability to export data to spreadsheet, e.g. Excel, LibreOffice etc.

For license information check the [LICENSE](LICENSE.md)-file.

[![Latest Stable Version](https://img.shields.io/packagist/v/yii2tech/spreadsheet.svg)](https://packagist.org/packages/yii2tech/spreadsheet)
[![Total Downloads](https://img.shields.io/packagist/dt/yii2tech/spreadsheet.svg)](https://packagist.org/packages/yii2tech/spreadsheet)
[![Build Status](https://travis-ci.org/yii2tech/spreadsheet.svg?branch=master)](https://travis-ci.org/yii2tech/spreadsheet)


Installation
------------

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist yii2tech/spreadsheet
```

or add

```json
"yii2tech/spreadsheet": "*"
```

to the require section of your composer.json.


Usage
-----

This extension provides ability to export data to a spreadsheet, e.g. Excel, LibreOffice etc.
It is powered by [phpoffice/phpspreadsheet](https://github.com/PHPOffice/PhpSpreadsheet) library.
Export is performed via `\yii2tech\spreadsheet\Spreadsheet` instance, which provides interface similar to `\yii\grid\GridView` widget.

Example:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ArrayDataProvider;

$exporter = new Spreadsheet([
    'dataProvider' => new ArrayDataProvider([
        'allModels' => [
            [
                'name' => 'some name',
                'price' => '9879',
            ],
            [
                'name' => 'name 2',
                'price' => '79',
            ],
        ],
    ]),
    'columns' => [
        [
            'attribute' => 'name',
            'contentOptions' => [
                'alignment' => [
                    'horizontal' => 'center',
                    'vertical' => 'center',
                ],
            ],
        ],
        [
            'attribute' => 'price',
        ],
    ],
]);
$exporter->save('/path/to/file.xls');
```

Please, refer to `\yii2tech\spreadsheet\Column` class for the information about column properties and configuration specifications.

While running web application you can use `\yii2tech\spreadsheet\Spreadsheet::send()` method to send a result file to
the browser through download dialog:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ActiveDataProvider;
use yii\web\Controller;

class ItemController extends Controller
{
    public function actionExport()
    {
        $exporter = new Spreadsheet([
            'dataProvider' => new ActiveDataProvider([
                'query' => Item::find(),
            ]),
        ]);
        return $exporter->send('items.xls');
    }
}
```


### Multiple sheet files <span id="multiple-sheet-files"></span>

You can create an output file with multiple worksheets (tabs). For example: you may want to export data about
equipment used in the office, keeping monitors, mouses, keyboards and so on in separated listings but in the same file.
To do so you will need to manually call `\yii2tech\spreadsheet\Spreadsheet::render()` method with different configuration
before creating final file. For example:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ActiveDataProvider;
use app\models\Equipment;

$exporter = (new Spreadsheet([
    'title' => 'Monitors',
    'dataProvider' => new ActiveDataProvider([
        'query' => Equipment::find()->andWhere(['group' => 'monitor']),
    ]),
    'columns' => [
        [
            'attribute' => 'name',
        ],
        [
            'attribute' => 'price',
        ],
    ],
]))->render(); // call `render()` to create a single worksheet

$exporter->configure([ // update spreadsheet configuration
    'title' => 'Mouses',
    'dataProvider' => new ActiveDataProvider([
        'query' => Equipment::find()->andWhere(['group' => 'mouse']),
    ]),
])->render(); // call `render()` to create a single worksheet

$exporter->configure([ // update spreadsheet configuration
    'title' => 'Keyboards',
    'dataProvider' => new ActiveDataProvider([
        'query' => Equipment::find()->andWhere(['group' => 'keyboard']),
    ]),
])->render(); // call `render()` to create a single worksheet

$exporter->save('/path/to/file.xls');
```

As the result you will get a single *.xls file with 3 worksheets (tabs): 'Monitors', 'Mouses' and 'Keyboards'.

Using `\yii2tech\spreadsheet\Spreadsheet::configure()` you can reset any spreadsheet parameter, including `columns`.
Thus you are able to combine several entirely different sheets into a single file.


### Large data processing <span id="large-data-processing"></span>

`\yii2tech\spreadsheet\Spreadsheet` allows exporting of the `\yii\data\DataProviderInterface` and `\yii\db\QueryInterface` instances.
Export is performed via batches, which allows processing of the large data without memory overflow.

In case of `\yii\data\DataProviderInterface` usage, data will be split to batches using pagination mechanism.
Thus you should setup pagination with page size in order to control batch size:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ActiveDataProvider;

$exporter = new Spreadsheet([
    'dataProvider' => new ActiveDataProvider([
        'query' => Item::find(),
        'pagination' => [
            'pageSize' => 100, // export batch size
        ],
    ]),
]);
$exporter->save('/path/to/file.xls');
```

> Note: if you disable pagination in your data provider - no batch processing will be performed.

In case of `\yii\db\QueryInterface` usage, `Spreadsheet` will attempt to use `batch()` method, if it is present in the query
class (for example in case `\yii\db\Query` or `\yii\db\ActiveQuery` usage). If `batch()` method is not available -
`yii\data\ActiveDataProvider` instance will be automatically created around given query.
You can control batch size via `\yii2tech\spreadsheet\Spreadsheet::$batchSize`:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ActiveDataProvider;

$exporter = new Spreadsheet([
    'query' => Item::find(),
    'batchSize' => 200, // export batch size
]);
$exporter->save('/path/to/file.xls');
```

> Note: despite batch data processing reduces amount of resources needed for spreadsheet file generation,
  your program may still easily end up with PHP memory limit error on large data. This happens because of
  large complexity of the created document, which is stored in the memory during the entire process.
  In case you need to export really large data set, consider doing so via simple CSV data format
  using [yii2tech/csv-grid](https://github.com/yii2tech/csv-grid) extension.


### Complex headers <span id="complex-headers"></span>

You may union some columns in the sheet header into a groups. For example: you may have 2 different data columns:
'Planned Revenue' and 'Actual Revenue'. In this case you may want to display them as a single column 'Revenue', split
into 2 sub columns: 'Planned' and 'Actual'.
This can be achieved using `\yii2tech\spreadsheet\Spreadsheet::$headerColumnUnions`. Its each entry
should specify 'offset', which determines the amount of columns to be skipped, and 'length', which determines
the amount of columns to be united. Other options of the union are the same as for regular column.
For example:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii\data\ArrayDataProvider;

$exporter = new Spreadsheet([
    'dataProvider' => new ArrayDataProvider([
        'allModels' => [
            [
                'column1' => '1.1',
                'column2' => '1.2',
                'column3' => '1.3',
                'column4' => '1.4',
                'column5' => '1.5',
                'column6' => '1.6',
                'column7' => '1.7',
            ],
            [
                'column1' => '2.1',
                'column2' => '2.2',
                'column3' => '2.3',
                'column4' => '2.4',
                'column5' => '2.5',
                'column6' => '2.6',
                'column7' => '2.7',
            ],
        ],
    ]),
    'headerColumnUnions' => [
        [
            'header' => 'Skip 1 column and group 2 next',
            'offset' => 1,
            'length' => 2,
        ],
        [
            'header' => 'Skip 2 columns and group 2 next',
            'offset' => 2,
            'length' => 2,
        ],
    ],
]);
$exporter->save('/path/to/file.xls');
```

> Note: only single level of header column unions is supported. You will need to deal with more complex
  cases on your own.


### Custom cell rendering <span id="custom-cell-rendering"></span>

Before `save()` or `send()` method is invoked, you are able to edit generated spreadsheet, making some
final adjustments to it. Several methods exist to facilitate this process:

 - `\yii2tech\spreadsheet\Spreadsheet::renderCell()` - renders specified cell with given content and style.
 - `\yii2tech\spreadsheet\Spreadsheet::applyCellStyle()` - applies specified style to the cell.
 - `\yii2tech\spreadsheet\Spreadsheet::mergeCells()` - merges sell range into single one.

You may use these methods, after document has been composed via `\yii2tech\spreadsheet\Spreadsheet::render()`,
to override or add some content. For example:

```php
<?php

use yii2tech\spreadsheet\Spreadsheet;
use yii2tech\spreadsheet\SerialColumn;
use yii\data\ArrayDataProvider;

$exporter = new Spreadsheet([
    'dataProvider' => new ArrayDataProvider([
        'allModels' => [
            [
                'id' => 1,
                'name' => 'first',
            ],
            [
                'id' => 2,
                'name' => 'second',
            ],
        ],
    ]),
    'columns' => [
        [
            'class' => SerialColumn::class,
        ],
        [
            'attribute' => 'id',
        ],
        [
            'attribute' => 'name',
        ],
    ],
])->render(); // render the document

// override serial column header :
$exporter->renderCell('A1', 'Overridden serial column header');

// add custom footer :
$exporter->renderCell('A4', 'Custom A4', [
    'font' => [
        'color' => [
            'rgb' => '#FF0000',
        ],
    ],
]);

// merge footer cells :
$exporter->mergeCells('A4:B4');

$exporter->save('/path/to/file.xls');
```

> Tip: you can use `\yii2tech\spreadsheet\Spreadsheet::$rowIndex` to get number of the row, which is next
  to the last rendered one.


CHANGELOG.md
Yii 2 Spreadsheet Data Export extension Change Log
==================================================

1.0.5, September 12, 2019
-------------------------

- Enh #23: Added `Spreadsheet::$writerCreator` allowing setup of custom spreadsheet writer (klimov-paul)


1.0.4, July 10, 2019
--------------------

- Enh #21: Added `Spreadsheet::$startRowIndex` allowing to skip lines at the sheet document beginning (klimov-paul)


1.0.3, January 24, 2019
-----------------------

- Bug #11: Fixed `Spreadsheet::send()` fails for 'Xlsx' writer (klimov-paul)


1.0.2, April 18, 2018
---------------------

- Bug #4: Fixed `Spreadsheet::save()` forces directory name to lowercase (klimov-paul)
- Bug #5: Fixed `Spreadsheet::createDataColumn()` sets column format to 'text' instead of 'raw' (klimov-paul)


1.0.1, April 9, 2018
--------------------

- Enh #2: `Spreadsheet::send()` now throws `\RuntimeException` in case temporary file can not be created (Eseperio, klimov-paul)


1.0.0, February 13, 2018
------------------------

- Initial release.


tests/TestCase.php
<?php

namespace yii2tech\tests\unit\spreadsheet;

use yii\helpers\ArrayHelper;
use Yii;
use yii\helpers\FileHelper;

/**
 * Base class for the test cases.
 */
class TestCase extends \PHPUnit\Framework\TestCase
{
    /**
     * {@inheritdoc}
     */
    protected function setUp()
    {
        parent::setUp();
        $this->mockApplication();
    }

    /**
     * {@inheritdoc}
     */
    protected function tearDown()
    {
        parent::tearDown();
        $this->destroyApplication();
        $this->removeTestFilePath();
    }

    /**
     * Populates Yii::$app with a new application
     * The application will be destroyed on tearDown() automatically.
     * @param array $config The application configuration, if needed
     * @param string $appClass name of the application class to create
     */
    protected function mockApplication($config = [], $appClass = '\yii\console\Application')
    {
        new $appClass(ArrayHelper::merge([
            'id' => 'testapp',
            'basePath' => __DIR__,
            'vendorPath' => $this->getVendorPath(),
            'components' => [
                'db' => [
                    'class' => 'yii\db\Connection',
                    'dsn' => 'sqlite::memory:',
                ],
            ],
        ], $config));
    }

    /**
     * @return string vendor path
     */
    protected function getVendorPath()
    {
        return dirname(__DIR__) . '/vendor';
    }

    /**
     * Destroys application in Yii::$app by setting it to null.
     */
    protected function destroyApplication()
    {
        Yii::$app = null;
    }

    /**
     * @return string test file path
     */
    protected function getTestFilePath()
    {
        return Yii::getAlias('@runtime/spreadsheet-test');
    }

    /**
     * Ensures test file path exists.
     * @return string test file path
     */
    protected function ensureTestFilePath()
    {
        $path = $this->getTestFilePath();
        FileHelper::createDirectory($path);
        return $path;
    }

    /**
     * Removes the test file path.
     */
    protected function removeTestFilePath()
    {
        $path = $this->getTestFilePath();
        FileHelper::removeDirectory($path);
    }

    /**
     * Invokes object method, even if it is private or protected.
     * @param object $object object.
     * @param string $method method name.
     * @param array $args method arguments
     * @return mixed method result
     */
    protected function invoke($object, $method, array $args = [])
    {
        $classReflection = new \ReflectionClass(get_class($object));
        $methodReflection = $classReflection->getMethod($method);
        $methodReflection->setAccessible(true);
        $result = $methodReflection->invokeArgs($object, $args);
        $methodReflection->setAccessible(false);
        return $result;
    }
}


tests/bootstrap.php
<?php

// ensure we get report on all possible php errors
error_reporting(-1);

define('YII_ENABLE_ERROR_HANDLER', false);
define('YII_DEBUG', true);
$_SERVER['SCRIPT_NAME'] = '/' . __DIR__;
$_SERVER['SCRIPT_FILENAME'] = __FILE__;

require_once(__DIR__ . '/../vendor/autoload.php');
require_once(__DIR__ . '/../vendor/yiisoft/yii2/Yii.php');

Yii::setAlias('@yii2tech/tests/unit/spreadsheet', __DIR__);
Yii::setAlias('@yii2tech/spreadsheet', dirname(__DIR__) . '/src');

require_once(__DIR__ . '/compatibility.php');

tests/ExcelGridTest.php
<?php

namespace yii2tech\tests\unit\spreadsheet;

use yii2tech\spreadsheet\ExcelGrid;
use Yii;
use yii\data\ArrayDataProvider;
use yii\i18n\Formatter;

class ExcelGridTest extends TestCase
{
    /**
     * Setup tables for test ActiveRecord
     */
    protected function setupTestDbData()
    {
        $db = Yii::$app->getDb();

        // Structure :

        $table = 'Item';
        $columns = [
            'id' => 'pk',
            'name' => 'string',
            'number' => 'integer',
        ];
        $db->createCommand()->createTable($table, $columns)->execute();

        $db->createCommand()->batchInsert($table, ['name', 'number'], [
            ['first', 1],
            ['second', 2],
            ['third', 3],
        ])->execute();
    }

    /**
     * @param array $config Excel grid configuration.
     * @return ExcelGrid Excel grid instance.
     */
    protected function createExcelGrid(array $config = [])
    {
        if (!isset($config['dataProvider']) && !isset($config['query'])) {
            $config['dataProvider'] = new ArrayDataProvider();
        }
        return new ExcelGrid($config);
    }

    // Tests :

    public function testSetupFormatter()
    {
        $grid = $this->createExcelGrid();

        $formatter = new Formatter();
        $grid->setFormatter($formatter);
        $this->assertSame($formatter, $grid->getFormatter());
    }

    public function testExport()
    {
        $grid = $this->createExcelGrid([
            'dataProvider' => new ArrayDataProvider([
                'allModels' => [
                    [
                        'id' => 1,
                        'name' => 'first',
                    ],
                    [
                        'id' => 2,
                        'name' => 'second',
                    ],
                ],
            ])
        ]);

        $fileName = $this->getTestFilePath() . '/save.xls';
        $grid->render()->save($fileName);

        $this->assertTrue(file_exists($fileName));
    }
}

src/DataColumn.php
<?php
/**
 * @link https://github.com/yii2tech
 * @copyright Copyright (c) 2015 Yii2tech
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yii2tech\spreadsheet;

use yii\base\Model;
use yii\data\ActiveDataProvider;
use yii\db\ActiveQueryInterface;
use yii\helpers\ArrayHelper;
use yii\helpers\Inflector;

/**
 * DataColumn is the default column type for the {@see Spreadsheet}.
 *
 * @author Paul Klimov <klimov.paul@gmail.com>
 * @since 1.0
 */
class DataColumn extends Column
{
    /**
     * @var string the attribute name associated with this column. When neither {@see Column::$content} nor {@see value}
     * is specified, the value of the specified attribute will be retrieved from each data model and displayed.
     *
     * Also, if {@see label} is not specified, the label associated with the attribute will be displayed.
     */
    public $attribute;
    /**
     * @var string label to be displayed in the {@see Column::$header} and also to be used as the sorting
     * link label when sorting is enabled for this column.
     * If it is not set and the models provided by the GridViews data provider are instances
     * of {@see \yii\db\ActiveRecord}, the label will be determined using {@see \yii\db\ActiveRecord::getAttributeLabel()}.
     * Otherwise {@see \yii\helpers\Inflector::camel2words()} will be used to get a label.
     */
    public $label;
    /**
     * @var string|\Closure an anonymous function or a string that is used to determine the value to display in the current column.
     *
     * If this is an anonymous function, it will be called for each row and the return value will be used as the value to
     * display for every data model. The signature of this function should be: `function ($model, $key, $index, $column)`.
     * Where `$model`, `$key`, and `$index` refer to the model, key and index of the row currently being rendered
     * and `$column` is a reference to the {@see DataColumn} object.
     *
     * You may also set this property to a string representing the attribute name to be displayed in this column.
     * This can be used when the attribute to be displayed is different from the {@see attribute} that is used for
     * sorting and filtering.
     *
     * If this is not set, `$model[$attribute]` will be used to obtain the value, where `$attribute` is the value of {@see attribute}.
     */
    public $value;
    /**
     * @var string|array in which format should the value of each data model be displayed as (e.g. `"raw"`, `"text"`, `"html"`,
     * `['date', 'php:Y-m-d']`). Supported formats are determined by the {@see Spreadsheet::$formatter} used by
     * the {@see Spreadsheet}. Default format is "raw" which will display value as it is.
     */
    public $format = 'raw';
    /**
     * @var string|array|bool the HTML code representing a filter input (e.g. a text field, a dropdown list)
     * that is used for this data column. This property is effective only when {@see Spreadsheet::$filterModel} is set.
     *
     * - If this property is not set, a text field will be generated as the filter input;
     * - If this property is an array, a dropdown list will be generated that uses this property value as
     *   the list options.
     * - If you don't want a filter for this data column, set this value to be false.
     */
    public $filter;


    /**
     * {@inheritdoc}
     */
    public function renderHeaderCellContent()
    {
        if ($this->header !== null || $this->label === null && $this->attribute === null) {
            return parent::renderHeaderCellContent();
        }

        $provider = $this->grid->dataProvider;

        if ($this->label === null) {
            if ($provider instanceof ActiveDataProvider && $provider->query instanceof ActiveQueryInterface) {
                /* @var $model Model */
                $model = new $provider->query->modelClass;
                $label = $model->getAttributeLabel($this->attribute);
            } else {
                $models = $provider->getModels();
                if (($model = reset($models)) instanceof Model) {
                    /* @var $model Model */
                    $label = $model->getAttributeLabel($this->attribute);
                } else {
                    $label = Inflector::camel2words($this->attribute);
                }
            }
        } else {
            $label = $this->label;
        }

        return $label;
    }

    /**
     * {@inheritdoc}
     */
    public function renderFilterCellContent()
    {
        if (is_string($this->filter)) {
            return $this->filter;
        }

        return parent::renderFilterCellContent();
    }

    /**
     * Returns the data cell value.
     * @param mixed $model the data model
     * @param mixed $key the key associated with the data model
     * @param int $index the zero-based index of the data model among the models array returned by {@see Spreadsheet::$dataProvider}.
     * @return string the data cell value
     */
    public function getDataCellValue($model, $key, $index)
    {
        if ($this->value !== null) {
            if (is_string($this->value)) {
                return ArrayHelper::getValue($model, $this->value);
            }
            return call_user_func($this->value, $model, $key, $index, $this);
        } elseif ($this->attribute !== null) {
            return ArrayHelper::getValue($model, $this->attribute);
        }

        return null;
    }

    /**
     * {@inheritdoc}
     */
    public function renderDataCellContent($model, $key, $index)
    {
        if ($this->content === null) {
            $value = $this->getDataCellValue($model, $key, $index);
            if ($value === null) {
                return $this->grid->nullDisplay;
            }
            return $this->grid->formatter->format($value, $this->format);
        }

        return parent::renderDataCellContent($model, $key, $index);
    }
}

src/ExcelGrid.php
<?php
/**
 * @link https://github.com/yii2tech
 * @copyright Copyright (c) 2015 Yii2tech
 * @license [New BSD License](http://www.opensource.org/licenses/bsd-license.php)
 */

namespace yii2tech\spreadsheet;

use yii\helpers\FileHelper;
use yii\i18n\Formatter;
use PHPExcel;
use PHPExcel_IOFactory;
use Yii;
use yii\base\Component;
use yii\base\InvalidConfigException;
use yii\di\Instance;

/**
 * ExcelGrid allows export of data provider into Excel document via [[PHPExcel]] library.
 * It provides interface, which is similar to [[\yii\grid\GridView]] widget.
 *
 * Example:
 *
 * ```php
 * $exporter = new ExcelGrid([
 *     'dataProvider' => new ArrayDataProvider([
 *         'allModels' => [
 *             [
 *                 'name' => 'some name',
 *                 'price' => '9879',
 *             ],
 *             [
 *                 'name' => 'name 2',
 *                 'price' => '79',
 *             ],
 *         ],
 *     ]),
 *     'columns' => [
 *         [
 *             'attribute' => 'name',
 *             'contentOptions' => [
 *                 'alignment' => [
 *                     'horizontal' => 'center',
 *                     'vertical' => 'center',
 *                 ],
 *             ],
 *         ],
 *         [
 *             'attribute' => 'price',
 *         ],
 *     ],
 * ]);
 * $exporter->render()->save('/path/to/file.xls');
 * ```
 *
 * @see http://phpexcel.codeplex.com/documentation
 * @see PHPExcel
 *
 * @property array|Formatter $formatter the formatter used to format model attribute values into displayable texts.
 *
 * @author Paul Klimov <klimov.paul@gmail.com>
 * @since 1.0
 */
class ExcelGrid extends Component
{
    /**
     * @var \yii\data\DataProviderInterface the data provider for the view. This property is required.
     */
    public $dataProvider;
    /**
     * @var array|Column[]
     */
    public $columns = [];
    /**
     * @var bool whether to show the header section of the sheet.
     */
    public $showHeader = true;
    /**
     * @var bool whether to show the footer section of the sheet.
     */
    public $showFooter = false;
    /**
     * @var string|null sheet title.
     */
    public $title;
    /**
     * @var string the HTML display when the content of a cell is empty.
     * This property is used to render cells that have no defined content,
     * e.g. empty footer or filter cells.
     *
     * Note that this is not used by the [[DataColumn]] if a data item is `null`. In that case
     * the [[nullDisplay]] property will be used to indicate an empty data value.
     */
    public $emptyCell = '';
    /**
     * @var string the text to be displayed when formatting a `null` data value.
     */
    public $nullDisplay = '';
    /**
     * @var string writer type (format type). If not set will be determined automatically.
     * Supported values:
     *
     * - 'Excel5'
     * - 'Excel2007'
     * - 'OpenDocument'
     * - 'CSV'
     * - 'PDF'
     * - 'HTML'
     */
    public $writerType;

    /**
     * @var int current sheet row index.
     */
    protected $rowIndex;

    /**
     * @var PHPExcel|null Excel document representation instance.
     */
    private $_document;
    /**
     * @var array|Formatter the formatter used to format model attribute values into displayable texts.
     * This can be either an instance of [[Formatter]] or an configuration array for creating the [[Formatter]]
     * instance. If this property is not set, the "formatter" application component will be used.
     */
    private $_formatter;


    /**
     * @return PHPExcel Excel document representation instance.
     */
    public function getDocument()
    {
        if (!is_object($this->_document)) {
            $this->_document = new PHPExcel();
        }
        return $this->_document;
    }

    /**
     * @param PHPExcel|null $document Excel document representation instance.
     */
    public function setDocument($document)
    {
        $this->_document = $document;
    }

    /**
     * @return Formatter formatter instance
     */
    public function getFormatter()
    {
        if (!is_object($this->_formatter)) {
            if ($this->_formatter === null) {
                $this->_formatter = Yii::$app->getFormatter();
            } else {
                $this->_formatter = Instance::ensure($this->_formatter, Formatter::className());
            }
        }
        return $this->_formatter;
    }

    /**
     * @param array|Formatter $formatter
     */
    public function setFormatter($formatter)
    {
        $this->_formatter = $formatter;
    }

    /**
     * Creates column objects and initializes them.
     */
    protected function initColumns()
    {
        if (empty($this->columns)) {
            $this->guessColumns();
        }
        foreach ($this->columns as $i => $column) {
            if (is_string($column)) {
                $column = $this->createDataColumn($column);
            } elseif (is_array($column)) {
                $column = Yii::createObject(array_merge([
                    'class' => DataColumn::className(),
                    'grid' => $this,
                ], $column));
            }
            if (!$column->visible) {
                unset($this->columns[$i]);
                continue;
            }
            $this->columns[$i] = $column;
        }
    }

    /**
     * This function tries to guess the columns to show from the given data
     * if [[columns]] are not explicitly specified.
     */
    protected function guessColumns()
    {
        $models = $this->dataProvider->getModels();
        $model = reset($models);
        if (is_array($model) || is_object($model)) {
            foreach ($model as $name => $value) {
                $this->columns[] = (string) $name;
            }
        }
    }

    /**
     * Creates a [[DataColumn]] object based on a string in the format of "attribute:format:label".
     * @param string $text the column specification string
     * @return DataColumn the column instance
     * @throws InvalidConfigException if the column specification is invalid
     */
    protected function createDataColumn($text)
    {
        if (!preg_match('/^([^:]+)(:(\w*))?(:(.*))?$/', $text, $matches)) {
            throw new InvalidConfigException('The column must be specified in the format of "attribute", "attribute:format" or "attribute:format:label"');
        }

        return Yii::createObject([
            'class' => DataColumn::className(),
            'grid' => $this,
            'attribute' => $matches[1],
            'format' => isset($matches[3]) ? $matches[3] : 'text',
            'label' => isset($matches[5]) ? $matches[5] : null,
        ]);
    }

    /**
     * @param array $properties list of document properties in format: name => value
     * @return $this self reference
     * @see \PHPExcel_DocumentProperties
     */
    public function properties($properties)
    {
        $documentProperties = $this->getDocument()->getProperties();
        foreach ($properties as $name => $value) {
            $method = 'set' . ucfirst($name);
            call_user_func([$documentProperties, $method], $value);
        }
        return $this;
    }

    /**
     * @param \yii\data\DataProviderInterface $dataProvider the data provider for the document
     * @return $this self reference
     */
    public function dataProvider($dataProvider)
    {
        $this->dataProvider = $dataProvider;
        return $this;
    }

    /**
     * Performs actual document composition.
     * @return $this self reference.
     */
    public function render()
    {
        $this->initColumns();

        $document = $this->getDocument();

        if ($this->rowIndex !== null) {
            // second run
            $document->createSheet();
            $document->setActiveSheetIndex($document->getActiveSheetIndex() + 1);
        }

        if ($this->title !== null) {
            $document->getActiveSheet()->setTitle($this->title);
        }

        $this->rowIndex = 1;

        $this->applyColumnOptions();

        if ($this->showHeader) {
            $this->renderHeader();
        }
        $this->renderBody();
        if ($this->showFooter) {
            $this->renderFooter();
        }

        return $this;
    }

    /**
     * Applies column overall options, such as dimension options.
     */
    protected function applyColumnOptions()
    {
        $sheet = $this->getDocument()->getActiveSheet();
        $columnIndex = 'A';
        foreach ($this->columns as $column) {
            /* @var $column Column */
            if (!empty($column->dimensionOptions)) {
                $columnDimension = $sheet->getColumnDimension($columnIndex);
                foreach ($column->dimensionOptions as $name => $value) {
                    $method = 'set' . ucfirst($name);
                    call_user_func([$columnDimension, $method], $value);
                }
            }

            $columnIndex++;
        }
    }

    /**
     * Renders sheet table header
     */
    protected function renderHeader()
    {
        $columnIndex = 'A';
        foreach ($this->columns as $column) {
            /* @var $column Column */
            $column->renderHeaderCell($columnIndex . $this->rowIndex);
            $columnIndex++;
        }
        $this->rowIndex++;
    }

    /**
     * Renders sheet table footer
     */
    protected function renderFooter()
    {
        $columnIndex = 'A';
        foreach ($this->columns as $column) {
            /* @var $column Column */
            $column->renderFooterCell($columnIndex . $this->rowIndex);
            $columnIndex++;
        }
        $this->rowIndex++;
    }

    /**
     * Renders sheet table body
     */
    protected function renderBody()
    {
        $models = array_values($this->dataProvider->getModels());
        $keys = $this->dataProvider->getKeys();
        foreach ($models as $index => $model) {
            $key = $keys[$index];
            $columnIndex = 'A';
            foreach ($this->columns as $column) {
                /* @var $column Column */
                $column->renderDataCell($columnIndex . $this->rowIndex, $model, $key, $index);
                $columnIndex++;
            }
            $this->rowIndex++;
        }
    }

    /**
     * Saves the document into a file.
     * @param string $filename name of the output file.
     */
    public function save($filename)
    {
        $filename = Yii::getAlias($filename);

        $writerType = $this->writerType;
        if ($writerType === null) {
            $fileExtension = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
            if ($fileExtension === 'xlsx') {
                $writerType = 'Excel2007';
            } else {
                $writerType = 'Excel5';
            }
        }

        $fileDir = strtolower(pathinfo($filename, PATHINFO_DIRNAME));
        FileHelper::createDirectory($fileDir);

        $objWriter = PHPExcel_IOFactory::createWriter($this->getDocument(), $writerType);
        $objWriter->save($filename);
    }

    /**
     * Sends the rendered content as a file to the browser.
     *
     * Note that this method only prepares the response for file sending. The file is not sent
     * until [[\yii\web\Response::send()]] is called explicitly or implicitly.
     * The latter is done after you return from a controller action.
     *
     * @param string $attachmentName the file name shown to the user.
     * @param array $options additional options for sending the file. The following options are supported:
     *
     *  - `mimeType`: the MIME type of the content. Defaults to 'application/octet-stream'.
     *  - `inline`: bool, whether the browser should open the file within the browser window. Defaults to false,
     *    meaning a download dialog will pop up.
     *
     * @return \yii\web\Response the response object
     */
    public function send($attachmentName, $options = [])
    {
        $tempFileName = tempnam(Yii::getAlias('@runtime'), 'ExcelGridTemp_');
        $fileExtension = strtolower(pathinfo($attachmentName, PATHINFO_EXTENSION));
        if (!empty($fileExtension)) {
            $tempFileName .= '.' . $fileExtension;
        }
        $this->save($tempFileName);
        $content = file_get_contents($tempFileName);
        $response = Yii::$app->getResponse()->sendContentAsFile($content, $attachmentName, $options);
        unlink($tempFileName);
        return $response;
    }
}